toolkit/tools/pkg/imagecustomizerlib/resolvconf.go

```diff
@@ -0,0 +1,142 @@
+func overrideResolvConf(imageChroot *safechroot.Chroot) (resolvConfInfo, error) {
+	logger.Log.Infof("Overriding resolv.conf file")
+
+	imageResolveConfPath := filepath.Join(imageChroot.RootDir(), resolvConfPath)
+
+	existing := resolvConfInfo{}
+
+	stat, err := os.Lstat(imageResolveConfPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return resolvConfInfo{}, fmt.Errorf("failed to stat resolv.conf file:\n%w", err)
+		}
+		existing.existingType = resolvConfTypeNone
+	} else if stat.Mode()&os.ModeSymlink != 0 {
+		symlinkPath, err := os.Readlink(imageResolveConfPath)
+		if err != nil {
+			return resolvConfInfo{}, fmt.Errorf("failed to read resolv.conf symlink's path:\n%w", err)
+		}
+		existing.existingType = resolvConfTypeSymlink
+		existing.symlinkPath = symlinkPath
+	} else {
+		fileContents, err := file.Read(imageResolveConfPath)
+		if err != nil {
+			return resolvConfInfo{}, fmt.Errorf("failed to read resolv.conf file:\n%w", err)
+		}
+		existing.existingType = resolvConfTypeFile
+		existing.fileContents = fileContents
+	}
+
+	// Remove the existing resolv.conf file, if it exists.
+	// Note: It is assumed that the image will have a process that runs on boot that will override the resolv.conf
+	// file. For example, systemd-resolved. So, it isn't neccessary to make a back-up of the existing file.
+	err = os.RemoveAll(imageResolveConfPath)
+	if err != nil {
+		return resolvConfInfo{}, fmt.Errorf("failed to delete existing resolv.conf file:\n%w", err)
+	}
+
+	err = file.Copy(resolvConfPath, imageResolveConfPath)
+	if err != nil {
+		return resolvConfInfo{}, fmt.Errorf("failed to override resolv.conf file with host's resolv.conf:\n%w", err)
+	}
+
+	return existing, nil
+}
+
+func restoreResolvConf(existing resolvConfInfo, imageChroot *safechroot.Chroot) error {
+	imageResolveConfPath := filepath.Join(imageChroot.RootDir(), resolvConfPath)
+
+	// Delete the overridden resolv.conf file.
+	err := os.RemoveAll(imageResolveConfPath)
+	if err != nil {
+		return fmt.Errorf("failed to delete overridden resolv.conf file: %w", err)
+	}
+
+	switch existing.existingType {
+	case resolvConfTypeNone:
+		// Check if systemd-resolved is enabled.
+		resolvedEnabled := true
+		err := imageChroot.UnsafeRun(func() error {
+			err := shell.ExecuteLive(true, "systemctl", "is-enabled", "systemd-resolved.service")
+
+			var exitError *exec.ExitError
+			if errors.As(err, &exitError) && exitError.ExitCode() == 1 {
+				resolvedEnabled = false
+				return nil
+			}
+			return err
+		})
+		if err != nil {
+			return fmt.Errorf("failed to check if systemd-resolved service is enabled:\n%w", err)
+		}
+
+		if resolvedEnabled {
+			// The systemd-resolved.service is enabled.
+			// So, create the symlink for the resolv.conf file.
+			// While technically this symlink will be automatically created on first boot, doing it now is helpful for
+			// verity images where the root filesystem is readonly.
+			err := os.Symlink(resolvSystemdStubPath, imageResolveConfPath)
```

Comment:
More for my own info: does it matter if the linked file actually exists? As in: should we inform/warn the user in case it's missing or it's controlled by systemd and even if it's not there now, it will get created eventually?

Reply:
For normal images, it is fine if the `resolv.conf` file is missing because `systemd-resolved` will create it on first-boot. But for verity images, `systemd-resolved` cannot create the file since the root filesystem is readonly. And this prevents networking from working.

Another approach would be to leave it up the user to create the `resolv.conf` file in their MIC config for verity images. But this adds extra complexity to the minimal verity config.

In general, for MIC, I generally try to follow the rule of not doing something unless the user explicitly asks for it. And this `resolv.conf` change is technically a violation of that rule. But I reason that since MIC needs to mess with the `resolv.conf` file anyway, it doesn't hurt to do a bit extra.